crates/sui-core/src/histogram.rs

```diff
@@ -0,0 +1,389 @@
+/// it exports each requested percentile into separate prometheus gauge, while original implementation creates
+/// gauge per bucket.
+/// It also exports _sum and _count aggregates same as original implementation.
+///
+/// It is ok to measure timings for things like network latencies and expensive crypto operations.
+/// However as a rule of thumb this histogram should not be used in places that can produce very high data point count.
+///
+/// As a last round of defence this histogram emits error log when too much data is flowing in and drops data points.
+///
+/// This implementation puts great deal of effort to make sure the metric does not cause any harm to the code itself:
+/// * Reporting data point is a non-blocking send to a channel
+/// * Data point collections tries to clear the channel as fast as possible
+/// * Expensive histogram calculations are done in a separate blocking tokio thread pool to avoid effects on main scheduler
+/// * If histogram data is produced too fast, the data is dropped and error! log is emitted
+impl HistogramVec {
+    pub fn new_in_registry(name: &str, desc: &str, labels: &[&str], registry: &Registry) -> Self {
+        Self::new_in_registry_with_percentiles(
+            name,
+            desc,
+            labels,
+            registry,
+            vec![500usize, 950, 990],
+        )
+    }
+
+    /// Allows to specify percentiles in 1/1000th, e.g. 90pct is specified as 900
+    pub fn new_in_registry_with_percentiles(
+        name: &str,
+        desc: &str,
+        labels: &[&str],
+        registry: &Registry,
+        percentiles: Vec<usize>,
+    ) -> Self {
+        let sum_name = format!("{}_sum", name);
+        let count_name = format!("{}_count", name);
+        let sum =
+            register_int_counter_vec_with_registry!(sum_name, desc, labels, registry).unwrap();
+        let count =
+            register_int_counter_vec_with_registry!(count_name, desc, labels, registry).unwrap();
+        let labels: Vec<_> = labels.iter().cloned().chain(["pct"].into_iter()).collect();
+        let gauge = register_int_gauge_vec_with_registry!(name, desc, &labels, registry).unwrap();
+        Self::new(gauge, sum, count, percentiles)
+    }
+
+    // Do not expose it to public interface because we need labels to have a specific format (e.g. add last label is "pct")
+    fn new(
+        gauge: IntGaugeVec,
+        sum: IntCounterVec,
+        count: IntCounterVec,
+        percentiles: Vec<usize>,
+    ) -> Self {
+        // The processing task is very fast and should not have realistic scenario where this channel overflows
+        // If the channel is growing too fast you will start seeing error! from the
+        // data points being dropped way before it will produce any measurable memory pressure
+        #[allow(clippy::disallowed_methods)]
+        let (sender, receiver) = mpsc::unbounded_channel();
```

Comment:
I'm curious why we chose an unbounded channel vs a bounded one with some high enough capacity? Agreed that we have a way to drop samples but why not have implicit drop on the client itself because we still buffer them in tokio channel before receiver gets a chance to drop them? We could potentially just run the loop and compute + update counters inline in one blocking task if we had a channel with bounded capacity

Reply:
Yeah, I should probably just switch to bounded channel